File: jl_src/day_13.jl
import ArgParse

# Build ArgParsing Table
s = ArgParse.ArgParseSettings()
@ArgParse.add_arg_table s begin
    "--input"
        help = "Filepath to input data"
        arg_type = String
        required = true
    "--verbose"
        help = "Add additional reporting"
        action = :store_true
end

abstract type Fold{T} end
struct FoldX{T} <: Fold{T}
    pos::T
end
struct FoldY{T} <: Fold{T}
    pos::T
end

# Parse arguments
parsed_args = ArgParse.parse_args(ARGS, s)

# Process files
struct ProcessFile{T} end
function ProcessFile{T}(filepath::String)::Tuple{Set{Tuple{T,T}},Vector{Fold{T}}} where T
    points = Set{Tuple{T,T}}()
    folds = Vector{Fold{T}}()
    open(filepath) do f
        for line in readlines(f)
            if line == ""
                continue
            end
            line_split = split(line)
            if line_split[1] == "fold"
                # We have a fold instruction
                fold_bits = split(line_split[3],"=")
                val = parse(T, fold_bits[2])
                if fold_bits[1] == "y"
                    push!(folds, FoldY{T}(val))
                else
                    push!(folds, FoldX{T}(val))
                end
            else
                num_bits = split(line, ',')
                push!(points, (parse(T,num_bits[1]), parse(T,num_bits[2])))
            end
        end
    end
    (points, folds)
end

function FoldPoints(points::Set{Tuple{T,T}}, fold::FoldX{T})::Set{Tuple{T,T}} where T
    new_points = Set{Tuple{T,T}}()
    for point in points
       y = point[2]
       x = point[1]
       if x > fold.pos
          x = 2*fold.pos-x
       end
       push!(new_points, (x,y))
    end
    new_points
end

function FoldPoints(points::Set{Tuple{T,T}}, fold::FoldY{T})::Set{Tuple{T,T}} where T
    new_points = Set{Tuple{T,T}}()
    for point in points
       y = point[2]
       x = point[1]
       if y > fold.pos
          y = 2*fold.pos-y
       end
       push!(new_points, (x,y))
    end
    new_points
end

function SolveProblem(points::Set{Tuple{T,T}}, folds::Vector{Fold{T}}) where T
    println("Day 13 problem 1: ", length(FoldPoints(points, folds[1])))

    # Do all folds
    for fold in folds
        points = FoldPoints(points, fold)
    end

    # Find maxes
    xs = collect(map(t -> t[1], collect(points)))
    max_x = maximum(xs)

    ys = collect(map(t -> t[2], collect(points)))
    max_y = maximum(ys)

    # Render images
    println("Day 13 problem 2")
    for y in 0:max_y
        line = ""
        for x in 0:max_x
            if (x,y) in points
                line = line*"#"
            else
                line = line*" "
            end
        end
        println(line)
    end
end

# Get input filepath
input_filepath = parsed_args["input"]
verbose = parsed_args["verbose"]

points, folds = ProcessFile{Int}(input_filepath)

SolveProblem(points, folds)

File: rust_src/src/day_13.rs
extern crate argparse;
extern crate num;

use std::io::{self, BufRead};
use std::fs::File;
use std::path::Path;
use std::collections::HashSet;
use std::hash::Hash;
use std::ops::{Add,Sub,Mul};
use std::str::FromStr;
use std::fmt::Debug;
use argparse::{ArgumentParser, StoreTrue, Store};

// Define readlines function discussed here:
// https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

#[derive(Debug,Hash,PartialEq,Eq)]
struct Point<T> {
    x: T,
    y: T,
}

impl<T> Point<T> {
    fn build(x: T, y:T) -> Point<T> {
        Point::<T> {x: x, y: y}
    }
}

enum Fold<T> {
    FoldX(T),
    FoldY(T),
}

fn process_input<P,T>(filename: P) -> (HashSet<Point<T>>, Vec<Fold<T>>)
where
    P: AsRef<Path>,
    T: Hash+PartialEq+Eq,
    T: Add+Add<Output=T>,
    T: Sub+Sub<Output=T>,
    T: FromStr,
    <T as FromStr>::Err: Debug
{
    let mut points: HashSet<Point<T>> = HashSet::new();
    let mut folds: Vec<Fold<T>> = Vec::new();
    // Reading the file
    if let Ok(lines) = read_lines(filename) {
        // Consumes the iterator, returns an (Optional) String
        for line in lines {
            if let Ok(ip) = line {
                if ip == "" {
                   continue
                }
                let tokens = ip.split_whitespace().collect::<Vec::<&str>>();
                if tokens[0] == "fold" {
                    let fold_vals: Vec<&str> = tokens[2].split('=').collect();
                    let fold_type = fold_vals[0];
                    let fold_val = fold_vals[1];
                    match fold_type {
                        "x" => {
                            folds.push(Fold::<T>::FoldX(fold_val.parse::<T>().unwrap()));
                        }
                        "y" => {
                            folds.push(Fold::<T>::FoldY(fold_val.parse::<T>().unwrap()));
                        }
                        _ => {
                            panic!("Encountered bad state!");
                        }
                    }
                } else {
                    let nums: Vec<&str> = ip.split(',').collect();
                    let x = nums[0].parse::<T>().unwrap();
                    let y = nums[1].parse::<T>().unwrap();
                    points.insert(Point::<T>::build(x,y));
                }
            }
        }
    } else {
		panic!("Problem opening file!");
	}
    (points, folds)
}

fn fold_point<T>(point: &Point<T>, fold: &Fold<T>) -> Point<T>
where
    T: Add+Add<Output=T>,
    T: Sub+Sub<Output=T>,
    T: Mul+Mul<Output=T>,
    T: num::NumCast,
    T: PartialOrd,
    T: Hash+PartialEq+Eq,
    T: Copy,
{
    match fold {
        Fold::<T>::FoldX(pos) => {
            let mut x = point.x;
            let y = point.y;
            if x > *pos {
                x = num::cast::<_,T>(2).unwrap()*(*pos)-x;
            }
            Point::<T>::build(x, y)
        },
        Fold::<T>::FoldY(pos) => {
            let x = point.x;
            let mut y = point.y;
            if y > *pos {
                y = num::cast::<_,T>(2).unwrap()*(*pos)-y;
            }
            Point::<T>::build(x, y)
        }
    }
}

fn fold_points<T>(points: &HashSet<Point<T>>, fold: &Fold<T>) -> HashSet<Point<T>>
where
    T: Add+Add<Output=T>,
    T: Sub+Sub<Output=T>,
    T: Mul+Mul<Output=T>,
    T: num::NumCast,
    T: PartialOrd,
    T: Hash+PartialEq+Eq,
    T: Copy,
{
    let mut new_points: HashSet<Point<T>> = HashSet::new();
    for p in points {
        new_points.insert(fold_point(p, fold));
    }
    new_points
}

fn solve_problem<T>(mut points: HashSet<Point<T>>, folds: &Vec<Fold<T>>)
where
    T: Add+Add<Output=T>,
    T: Sub+Sub<Output=T>,
    T: Mul+Mul<Output=T>,
    T: num::NumCast,
    T: PartialOrd,
    T: Hash+PartialEq+Eq,
    T: Copy,
    T: Ord,
    T: num::One,
{
    println!("Day 13 problem 1: {}", fold_points(&points, &folds[0]).len());

    // Do all folds
    for fold in folds {
        points = fold_points(&points, fold);
    }

    let max_x: T = points.iter().map(|p| p.x).max().unwrap();
    let max_y: T = points.iter().map(|p| p.y).max().unwrap();

    println!("Day 13 problem 2:");

    for y in num::range_inclusive::<T>(num::cast::<_,T>(0).unwrap(),max_y) {
        let mut line = String::new();
        for x in num::range_inclusive::<T>(num::cast::<_,T>(0).unwrap(), max_x) {
            let p = Point::<T>::build(x,y);
            if points.contains(&p) {
                line = line+"#";
            } else {
                line = line+" ";
            }
        }
        println!("{}", line);
    }
}

fn main() {
    // Argument Parsing
    let mut verbose = false;
    let mut input = String::new();
    {
        let mut ap = ArgumentParser::new();
        ap.set_description("Day 13 AOC 2021");
        ap.refer(&mut input)
            .add_option(&["-i", "--input"], Store,
            "Filepath to input data");
        ap.refer(&mut verbose)
            .add_option(&["-v", "--verbose"], StoreTrue,
            "Add additional reporting");
        ap.parse_args_or_exit();
    }

    let (points, folds) = process_input::<_,i32>(input);

    solve_problem(points, &folds);
}
